[RFC 0123] Flake names (#123)
NixOS/rfcs: rfcs/0123-flake-names.md

---
feature: flake-names
start-date: 2022-03-12
author: Anselm Schüler
co-authors: None
shepherd-team: None
shepherd-leader: None
related-issues: None
---

# Summary
[summary]: #summary

Flakes can declare the field `name`.  
It represents the name of the flake.  
The derivations for a flake are no longer called `source`, but use the flake name.

# Motivation
[motivation]: #motivation

Flake-centric workflows often end up with a lot of derivations named “source”, and it’s difficult to navigate this.
Also, the discoverability and usability of flakes needs to be improved. Current commands mostly show technical information. This would be a step in the right direction.

# Detailed design
[design]: #detailed-design

A new supported property for flakes is introduced, `name`.  
Running `nix flake metadata` on a flake that declares this field displays it at the top.  
The derivation that contains the flake’s content is called `flake-source-${name}` or, if a short revision identifier is available, `flake-source-${name}-${shortRev}`.

# Examples and Interactions
[examples-and-interactions]: #examples-and-interactions

> A flake that is in the registry should have its name match the registry identifier.

Reviewer: Wouldn't that prevent the use of alternate names, or having multiple versions of the same flake in the registry?
(I use both)

Author: That’s a good point

Author: Maybe it’d be a good rule for the global registry, though, but it’d probably be followed anyways, and would only hinder edge-cases

---
feature: flake-names
start-date: 2022-03-12
author: Anselm Schüler
co-authors: None
shepherd-team: None
shepherd-leader: None
related-issues: None
---

# Summary
[summary]: #summary

Flakes can declare the field `name`.  
It represents the name of the flake.  

> The derivations for a flake are no longer called `source`, but use the flake name.

Reviewer 2: ```suggestion
The store paths corresponding to a flake source are no longer called `source`, but use the flake name.
```

(I’m nitpicking a bit here, but technically these aren’t derivations, they are just paths that are copied directly to the store)

Author: Good point!

# Motivation
[motivation]: #motivation

> Flake-centric workflows often end up with a lot of derivations named “source”, and it’s difficult to navigate this.

Reviewer 2: That’s probably worth a bit more explanations: for most intents and purposes, the store is supposed to be an opaque thing, and not meant to be directly “navigated”.

Now it is a very leaky abstraction, so my remark probably doesn’t really hold in practice, but it would be good to make this explicit (maybe show some concrete examples of where having all the sources be called “source” is an issue)

Author: It's an issue if you want to know what a nix command is doing at the moment. It often just displays “building XXXX-source”

Reviewer 2: I might be wrong, but that exact message shouldn’t happen (not because of flakes at least). This message happens when Nix builds a derivation, but flake sources aren’t derivations, they are directly instantiated during the evaluation (with `builtins.fetchTree`) using a different code path with amongst other things a more informative display.

What you might get though is something like “fetching source from https://cache.nixos.org” (because although they aren’t derivations, they can be substituted from the binary cache). Probably we could improve the message in that case.

Author: Oh, that’s good to know, sorry for not knowing that.

Author: About the store path being meant to be opaque: In my opinion that’s just practically not true at all, and the fact that derivation names exist at all proves it. If anything, this is useful for people that want to spelunk in the depths of what Nix does.

I also regularly use store paths to figure out what package or other source provides some binary in my profile. I have a shell function `goto-bin` for going into the directory of the real path of an executable, and I sometimes skim `realpath` output to see a derivation name.

This can also make flake-centric commands friendlier and easier to approach.

# Detailed design
[design]: #detailed-design

A new supported property for flakes is introduced, `name`.  
The derivation that contains the flake’s content is called `flake-source-${name}`  

> If a source control tag (e.g. git tag such as V3) is available, it is `flake-source-${name}-${tag}`.

Reviewer 2: This isn’t really possible without breaking purity afaik: The same git commit can have or not a tag attached to it (since tags are mutable), and evaluating a flake from that commit with or without the tag would yield a different result

Author: I thought I had removed this section? Someone already raised this issue.
Yes, that's a good point.

# Examples and Interactions
[examples-and-interactions]: #examples-and-interactions

Running `nix flake metadata` on a flake that declares this field displays it at the top.  
Running `nix flake show` on a flake that declares this field shows the name instead of the URL, followed by the URL in parentheses.

Examples:

File `/example/flake.nix`
```nix
{
  name = "example";
  outputs = { ... }: {
    lib.example = "example";
  };
}
```

Shell
```console
$ nix eval git+file:///example
[…] copying flake-source-example
"example"
```

Example of interactions:

Shell (using the previous file)
```
$ nix flake metadata /example
Name:          example
Resolved URL:  git+file:///example
Locked URL:    …
$ nix flake show /example
example (git+file:///home/anselmschueler/Code/example?rev=b0&rev=c714c8624f5d49a9d88e6e24550dd88515923c18)
└───lib: unknown
```

> # Drawbacks

Reviewer 2: The big drawback here (where “big” would probably need to be refined, I’ve no idea how much of an issue it is in practice) is that since the name is part of the store path, two flakes with the same content but a different name will end-up in a different place in the store − which itself means that everything that depends on them would have to be rebuild.

Author: But flakes with a different name will necessarily  have different content? Or are the flake.* files not copied to the store?

Reviewer 2: Fair point (I had a convoluted scenario in mind, but that’s probably way out of the scope of this RFC now that I think of it). 

That’s actually highlighting an interesting design challenge: There’s currently a neat separation between the fetching code and the code that knows about what the flake is. `inputs.foo = something` gets translated to a call to `builtins.fetchTree something`, and its result is then fed to the `outputs` function. And the `fetchTree` part doesn’t know (nor has to know) that it’s fetching a flake.
But if the name gets defined in the flake (and used in the output path), then the `fetchTree` logic must be a bit more convoluted:
- It has to know whether it is fetching a flake
- When fetching a flake, it must get its `flake.nix`, evaluate it to find its `name`, and only then put it in the store with the correct `name`.

Author: Wouldn’t it be better to have this conversation outside the review widget? I feel like this aspect of the discussion is somewhat hidden & non-chronological.

Reviewer 3: > * When fetching a flake, it must get its `flake.nix`, evaluate it to find its `name`, and only then put it in the store with the correct `name`.

This can be solved by adding `name` to the lock file.
We can use the same trick for other fetch-related parameters, like whether to fetch submodules or not, whether to git-crypt decrypt, etc, all of which would also be nice to configure in `flake.nix`.
`fetchTree` remains unaware of flake-ness.

---
feature: flake-names
start-date: 2022-03-12
author: Anselm Schüler
co-authors: None
shepherd-team: None
shepherd-leader: None
related-issues: None
---

# Summary
[summary]: #summary

Flakes can declare the field `name`.  
It represents the name of the flake.  
The store paths corresponding to a flake source are no longer called `source`, but use the flake name.

# Motivation
[motivation]: #motivation

- Flakes generate store paths named “source”, and it’s difficult to navigate this when manually inspecting the store.
- This metadata can be used to make flakes more approachable and usable, in particular, it can be output instead or with the URI, making the nix commands friendlier.

> 

Reviewer 3: ```suggestion
- Unique names are useful for the purpose of finding duplicate things caused by diamond dependencies. (E.g. `a -> b -> d1`, `a -> c -> d2`. Without a unique name (`d`), it is not possible to correlate `d1` and `d2` reliably.)

```

 - Most relevant when https://github.com/NixOS/nix/issues/5570 remains unimplemented. "Userspace" experimentation may yield some insights into solutions for 5570.

The derivation that contains the flake’s content is called `flake-source-${name}`  

# Examples and Interactions
[examples-and-interactions]: #examples-and-interactions

Running `nix flake metadata` on a flake that declares this field displays it at the top.  
Running `nix flake show` on a flake that declares this field shows the name instead of the URL, followed by the URL in parentheses.

Examples:

File `/example/flake.nix`
```nix
{
  name = "example";
  outputs = { ... }: {
    lib.example = "example";
  };
}
```

Shell
```console
$ nix eval git+file:///example
[…] copying flake-source-example
"example"
```

Example of interactions:

Shell (using the previous file)
```
$ nix flake metadata /example
Name:          example
Resolved URL:  git+file:///example
Locked URL:    …
$ nix flake show /example
example (git+file:///home/anselmschueler/Code/example?rev=b0&rev=c714c8624f5d49a9d88e6e24550dd88515923c18)
└───lib: unknown
```

# Drawbacks
[drawbacks]: #drawbacks

- This may cause clutter and additional maintenance.
- Since this changes the output of nix flake metadata and nix flake show, it might cause scripts that read this output to break.

> - This requires a significant change to the way flakes are handled.

Reviewer 3: ```suggestion
```

All changes come with an implementation cost. I don't think this RFC is special, because it seems like an addition that doesn't require action by the ecosystem, especially if the name is allowed to be unset.

> - This treats the Nix store as a user-facing part of Nix, which is generally not intended.

Reviewer 3: ```suggestion
```

The store has always been user-facing. Users interact with it, store paths show up in logs, etc.

Sometimes you don't have to think much about the store and that's good.

How will users will have to pay more attention to the store after this change?
